wagtail/blocks/list_block.py

```diff
@@ -24,25 +24,61 @@
 
 class ListBlockValidationError(ValidationError):
     def __init__(self, block_errors=None, non_block_errors=None):
-        self.non_block_errors = non_block_errors or ErrorList()
-        self.block_errors = block_errors or []
+        # non_block_errors may be passed here as an ErrorList, a plain list (of strings or
+        # ValidationErrors), or None.
+        # Normalise it to be an ErrorList, which provides an as_data() method that consistently
+        # returns a flat list of ValidationError objects.
+        # (note: iterating over ErrorList itself appears to give a list of message strings,
+        # but doesn't correctly account for ValidationErrors containing multiple messages)
+        # (note 2: items in this list are expected to be plain ValidationError instances; there is
+        # no special treatment of subclasses such as StructBlockValidationError)
+        self.non_block_errors = ErrorList(non_block_errors)
+
+        # block_errors may be passed here as a list (corresponding to the block value's elements,
+        # with None for child blocks with no errors) or a dict whose keys are the indexes of the
+        # child blocks with errors.
```

Comment:
If dict is the preferred format, maybe we should mention it first?

```suggestion
        # block_errors may be passed here as a dict whose keys are the indexes of the
        # child blocks with errors or a list (corresponding to the block value's elements,
        # with None for child blocks with no errors).
```